File: frontend/src/app/core/services/pending-action.service.ts
import { Injectable, inject } from '@angular/core';
import { ApiService } from './api.service';
import { ApprovalStatus, PendingAction, PendingActionPage } from '../../models/pending-action.model';

/**
 * Supervised-mode approval inbox (#3a): list / approve / reject actions parked for human review,
 * and the pending count for the nav badge.
 */
@Injectable({ providedIn: 'root' })
export class PendingActionService {
  private readonly api = inject(ApiService);
  private readonly basePath = '/pending-actions';

  list(status?: ApprovalStatus) {
    return this.api.get<PendingActionPage>(`${this.basePath}${status ? `?status=${status}` : ''}`);
  }

  count() {
    return this.api.get<{ pending: number }>(`${this.basePath}/count`);
  }

  approve(id: string) {
    return this.api.post<PendingAction>(`${this.basePath}/${id}/approve`, {});
  }

  reject(id: string, note?: string) {
    const q = note ? `?note=${encodeURIComponent(note)}` : '';
    return this.api.post<PendingAction>(`${this.basePath}/${id}/reject${q}`, {});
  }

  /** #3c: teach the correct category from this action's email and reject the wrongly-triggered action. */
  reclassify(id: string, categoryId: string) {
    return this.api.post<PendingAction>(`${this.basePath}/${id}/reclassify?categoryId=${categoryId}`, {});
  }
}

File: backend/src/main/java/com/postwerk/service/PendingActionService.java
package com.postwerk.service;

import com.postwerk.dto.automation.PendingActionResponse;
import com.postwerk.model.AutomationNode;
import com.postwerk.model.enums.ApprovalStatus;
import com.postwerk.service.executor.ExecutionContext;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.Map;
import java.util.UUID;

/**
 * Supervised mode (#3a): manages {@code PendingAction}s — side effects parked for human approval
 * instead of being performed during a live automation run.
 *
 * @since 1.0
 */
public interface PendingActionService {

    /**
     * Parks an action node's already-resolved side effect for approval. Called by the executor when
     * a live run reaches an {@code ACTION} node whose {@code executionMode} is {@code REVIEW}.
     *
     * @param node    the action node
     * @param detail  the resolved payload (the detail map the dry-run pass produced)
     * @param context the live execution context (source email/account)
     * @param userId  the automation owner
     */
    void park(AutomationNode node, Map<String, Object> detail, ExecutionContext context, UUID userId);

    /** The organization's approval inbox, optionally filtered by status (null = all), newest first. */
    Page<PendingActionResponse> list(UUID organizationId, ApprovalStatus status, Pageable pageable);

    /** Count of the organization's actions still awaiting a decision. */
    long countPending(UUID organizationId);

    /** Approves a pending action and performs its parked side effect ({@code deciderUserId} = approver). */
    PendingActionResponse approve(UUID organizationId, UUID deciderUserId, UUID id);

    /** Rejects a pending action; its side effect is never performed. */
    PendingActionResponse reject(UUID organizationId, UUID deciderUserId, UUID id, String note);

    /**
     * #3c: corrects the categorization that drove this action — teaches {@code correctCategoryId} the
     * action's source email (as a learning example, re-embedding the category) and rejects the action.
     */
    PendingActionResponse reclassify(UUID organizationId, UUID userId, UUID id, UUID correctCategoryId);
}
